posteriors/ekf/dense_fisher.py

```diff
@@ -0,0 +1,198 @@
+        aux: Auxiliary information from the log_likelihood call.
+    """
+
+    params: TensorTree
+    cov: torch.Tensor
+    log_likelihood: float = 0
+    aux: Any = None
+
+
+def init(
+    params: TensorTree,
+    init_cov: torch.Tensor | float = 1.0,
+) -> EKFDenseState:
+    """Initialise Multivariate Normal distribution over parameters.
+
+    Args:
+        params: Initial mean of the Normal distribution.
+        init_cov: Initial covariance matrix of the Multivariate Normal distribution.
+            If it is a float, it is defined as an identity matrix scaled by that float.
+
+    Returns:
+        Initial EKFDenseState.
+    """
+    if is_scalar(init_cov):
+        num_params = tree_size(params)
+        init_cov = init_cov * torch.eye(num_params, requires_grad=False)
+
+    return EKFDenseState(params, init_cov)
+
+
+def update(
+    state: EKFDenseState,
+    batch: Any,
+    log_likelihood: LogProbFn,
+    lr: float,
+    transition_cov: torch.Tensor | float = 0.0,
+    per_sample: bool = False,
+    inplace: bool = False,
+) -> EKFDenseState:
+    """Applies an extended Kalman Filter update to the Multivariate Normal distribution.
+    The approximate Bayesian update is based on the linearization
+    $$
+    \\log p(θ | y) ≈ \\log p(θ) +  ε g(μ)ᵀ(θ - μ) +  \\frac12 ε (θ - μ)^T F(μ) (θ - μ)
+    $$
+    where $μ$ is the mean of the prior distribution, $ε$ is the learning rate
+    (or equivalently the likelihood inverse temperature),
+    $g(μ)$ is the gradient of the log likelihood at μ and $F(μ)$ is the
+    empirical Fisher information matrix at $μ$ for data $y$.
+
+    Args:
+        state: Current state.
+        batch: Input data to log_likelihood.
+        log_likelihood: Function that takes parameters and input batch and
+            returns the log-likelihood value as well as auxiliary information,
+            e.g. from the model call.
+        lr: Inverse temperature of the update, which behaves like a learning rate.
+        transition_cov: Covariance of the transition noise, to additively
+            inflate the covariance before the update.
+        per_sample: If True, then log_likelihood is assumed to return a vector of
+            log likelihoods for each sample in the batch. If False, then log_likelihood
+            is assumed to return a scalar log likelihood for the whole batch, in this
+            case torch.func.vmap will be called, this is typically slower than
+            directly writing log_likelihood to be per sample.
+        inplace: Whether to update the state parameters in-place.
+
+    Returns:
+        Updated EKFDenseState.
+    """
+    if not per_sample:
+        log_likelihood = per_samplify(log_likelihood)
+
+    with torch.no_grad(), CatchAuxError():
+        log_liks, aux = log_likelihood(state.params, batch)
+        jac, _ = jacrev(log_likelihood, has_aux=True)(state.params, batch)
+        grad = tree_map(lambda x: x.mean(0), jac)
```

Comment:
```suggestion
        def log_likelihood_reduced(params, batch):
            per_samp_log_lik, internal_aux = log_likelihood(params, batch)
            return per_samp_log_lik.mean(), internal_aux

        grad, (log_liks, aux) = grad_and_value(log_likelihood_reduced, has_aux=True)(
            state.params, batch
        )
```

Calls to `jacrev` are [significantly more expensive](https://github.com/google/jax/pull/762#issuecomment-1001885437) than `grad`, so we should avoid calling it twice here (there is the internal call to `jacrev` inside `empirical_fisher`).

What do you think of my solution above? (Imports will likely need cleaning up and `from torch.func import grad_and_value` added)

Reply:
Looks good, thanks! Added the import.